### PR TITLE
Wrap Calcium fields/numbers

### PR DESCRIPTION
**Pull request by fredrik-johansson**
Follows #1070.

This adds CalciumField and the ca type.

Work in progress; has most of the functionality (sans ca_poly and ca_mat) but needs docs and tests.

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
Already a test failure, though only with Julia 1.3:

    Run julia-actions/julia-runtest@latest
       Testing Nemo
     Resolving package versions...
    Test Summary:   | Pass  Total
    ca.constructors |   18     18
    Test Summary:     | Pass  Total
    fmpq.constructors |   27     27

    signal (11): Segmentation fault
    in expression starting at /home/runner/work/Nemo.jl/Nemo.jl/test/flint/fmpq-test.jl:57
    ca_clear at /workspace/srcdir/calcium-0.4.0/ca/clear.c:80
    _ca_clear_fn at /home/runner/work/Nemo.jl/Nemo.jl/src/calcium/CalciumTypes.jl:191
    _jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
   ...

Looks like finalizers not being called in order, as I worried about.

**Comment by fredrik-johansson** [comment]
1.0 and 1.1 tests are apparently failing because they don't get the updated Calcium_jll. How to fix?

If 1.3 is the only version where ca_clear crashes, it suggests to me that our code is fine but Julia 1.3 has a GC bug. If so, we could omit the finalizer for CalciumField when version == 1.3. This would leak memory but otherwise would be harmless.

**Comment by thofma** [comment]
Ah, I need to update the pre 1.3 binaries.

Regarding 1.3 and a possible GC bug. Can you explain why there is a bug? They changed the order of the finalizers between julia versions at some point. Since this is a implementation detail and might change again, I would not rely on it.

**Comment by fredrik-johansson** [comment]
C = CalciumField()        # holds a ca_ctx_t
    x = C()                   # holds ca_t, plus a reference to C (x.parent)

The finalizer for `C` calls `ca_ctx_clear`.

The finalizer for `x` calls `ca_clear` with x and `x.parent` as arguments. This requires that `C` is still alive and well, i.e. that the Julia object `C` is alive and that its finalizer has not been called.

Everything should be fine if Julia calls the finalizer for every `ca` instance with `x.parent == C` before running the finalizer for `C`. You would hope that the `x.parent` reference ensures this.

**Comment by fredrik-johansson** [comment]
As discussed in #1064, this is potentially a problem for Antic number fields and maybe some of the Flint rings as well.

**Comment by thofma** [comment]
> Everything should be fine if Julia calls the finalizer for every `ca` instance with `x.parent == C` before running the finalizer for `C`. You would hope that the `x.parent` reference ensures this.

Since I expect julia to be able to cleanup recursive objects, I don't understand why this should be expected. I don't want the GC to prove at runtime that things are not recursive, so I don't expect too much when it comes to the order of the finalized calls.

**Comment by fredrik-johansson** [comment]
Unfortunately, I can't reproduce the crash with 1.3 on my own computer.

It's also a mystery that Antic nf_elems haven't failed so far :-)

**Comment by thofma** [comment]
For some reason I cannot push directly to your branch, so you need to merge https://github.com/fredrik-johansson/Nemo.jl/pull/2

**Comment by fredrik-johansson** [comment]
Manual reference counting seems to be working.

**Comment by fredrik-johansson** [comment]
Any ideas why [matching: [Oscar#master,Hecke#master,Singular#master] - ubuntu-latest, julia ~1.6.0-0] is stuck in "Run tests" for two hours with *no output at all*?

**Comment by thofma** [comment]
I have restarted it.

**Comment by fredrik-johansson** [comment]
@thofma  Do you have a workaround for https://github.com/JuliaDocs/Documenter.jl/issues/558 ?

I want to document e.g. (R::AcbField)(a::ca; parts::Bool=false).

Or should I make a named function for this?

**Comment by thofma** [comment]
No, there is no work around as far as I know. We usually handcraft the "constructor" section like here: https://nemocas.github.io/AbstractAlgebra.jl/latest/polynomial_rings/#Constructors.

**Comment by fredrik-johansson** [comment]
This is ready for review!

**Comment by fredrik-johansson** [comment]
The rand function is a placeholder, by the way. I can open a separate issue to discuss what kind of options we want here.

File: src/calcium/ca.jl [hunk 1]
```diff
@@ -0,0 +1,1447 @@
...
+   ccall((:ca_euler, libcalcium), Nothing, (Ref{ca}, Ref{CalciumField}), r, C)
+   return r
+end
+
+@doc Markdown.doc"""
+    const_i(C::CalciumField)
+
+Return the imaginary unit $i$ as an element of *C*.
+"""
+function const_i(C::CalciumField)
```
Comment (thofma) [inline src/calcium/ca.jl:681]: For `AcbField` this is called `onei`.

**Review by thofma — COMMENTED** [review]
In the docstring and in the docs, you are using `*a*` quite a alot. I think all of these should be `` `a` ``? I assume they should be formated as code and not italicized?

File: src/calcium/CalciumTypes.jl [hunk 2]
```diff
@@ -76,4 +96,111 @@ function _qqbar_clear_fn(a::qqbar)
...
+   end
+
+end
+
+function _ca_clear_fn(a::ca)
+   ccall((:ca_clear, libcalcium),
+        Nothing, (Ref{ca}, Ref{CalciumField}), a, a.parent)
+   a.parent.refcount -= 1
+   if a.parent.refcount == 0
+      ccall((:ca_ctx_clear, libcalcium), Nothing, (Ref{CalciumField},), a.parent)
```
Comment (wbhart) [inline src/calcium/CalciumTypes.jl:203]: Codecov seems to indicate ca_ctx_clear is never called. Could it be the refcounting is not working?

File: src/calcium/ca.jl [hunk 3]
```diff
@@ -0,0 +1,1447 @@
...
+#
+###############################################################################
+
+parent(a::ca) = a.parent
+
+parent_type(::Type{ca}) = CalciumField
+
+elem_type(::Type{CalciumField}) = ca
+
+base_ring(a::CalciumField) = Union{}   #  ?
```
Comment (wbhart) [inline src/calcium/ca.jl:24]: I think these are correct. No need for the ?

File: src/calcium/ca.jl [hunk 4]
```diff
@@ -0,0 +1,1447 @@
+###############################################################################
+#
+#   ca.jl : Calcium field elements
+#
+###############################################################################
+
+export ca, CalciumField, infinity, unsigned_infinity, undefined, unknown,
```
Comment (wbhart) [inline src/calcium/ca.jl:7]: Alphabetical order please. I know there are no other good examples of this.

File: src/calcium/ca.jl [hunk 5]
```diff
@@ -0,0 +1,1447 @@
...
+   return r
+end
+
+###############################################################################
+#
+#   Hashing
+#
+###############################################################################
+
+# todo: implement nontrivial hash functions on C
```
Comment (wbhart) [inline src/calcium/ca.jl:44]: Does this await something on the C side?

File: src/calcium/ca.jl [hunk 6]
```diff
@@ -0,0 +1,1447 @@
...
+function +(a::ca, b::Int)
+   C = a.parent
+   r = C()
+   ccall((:ca_add_si, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Int, Ref{CalciumField}), r, a, b, C)
+   check_special(r)
+   return r
+end
+
+function +(a::ca, b::fmpz)
```
Comment (wbhart) [inline src/calcium/ca.jl:184]: Do we have generic code that switches the order of arguments if one ad hoc operation is available?

File: src/calcium/ca.jl [hunk 7]
```diff
@@ -0,0 +1,1447 @@
...
+function //(a::ca, b::ca)
+   a, b = same_parent(a, b)
+   C = a.parent
+   r = C()
+   ccall((:ca_div, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Ref{ca}, Ref{CalciumField}), r, a, b, C)
+   check_special(r)
+   return r
+end
+
```
Comment (wbhart) [inline src/calcium/ca.jl:326]: Could you separate into sections Ad hoc division, Powering, Unary operations, for conformity with the other files.

File: src/calcium/ca.jl [hunk 8]
```diff
@@ -0,0 +1,1447 @@
...
+function inv(a::ca)
+   C = a.parent
+   r = C()
+   ccall((:ca_inv, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Ref{CalciumField}), r, a, C)
+   check_special(r)
+   return r
+end
+
+###############################################################################
```
Comment (wbhart) [inline src/calcium/ca.jl:457]: We usually put the predicates before the arithmetic ops, usually in a section called Basic manipulation.

File: src/calcium/ca.jl [hunk 9]
```diff
@@ -0,0 +1,1447 @@
...
+
+function zero!(z::ca)
+   C = z.parent
+   ccall((:ca_zero, libcalcium), Nothing, (Ref{ca}, Ref{CalciumField}), z, C)
+   return z
+end
+
+function mul!(z::ca, x::ca, y::ca)
+   if z.parent != x.parent || x.parent != y.parent
```
Comment (wbhart) [inline src/calcium/ca.jl:1351]: We do not check parents in unsafe operators.

File: src/calcium/qqbar.jl [hunk 10]
```diff
@@ -770,7 +782,7 @@ end
 Return the extension of the real sign function taking the value 1
 strictly in the right half plane, -1 strictly in the left half plane,
 and the sign of the imaginary part when on the imaginary axis.
-Equivalently, $\csgn(x) = x / \sqrt{x^2}$ except that the value is 0
+Equivalently, $\operatorname{csgn}(x) = x / \sqrt{x^2}$ except that the value is 0
```
Comment (wbhart) [inline src/calcium/qqbar.jl:785]: is \operatorname standard latex?

File: test/calcium/ca-test.jl [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+   @test !is_signed_inf(C(1))
+   @test !isundefined(C(1))
+   @test !isunknown(C(1))
+
+   @test und == und
+   @test_throws ErrorException (unk == unk)
+
+end
+
+@testset "ca.adhoc-operations" begin
```
Comment (wbhart) [inline test/calcium/ca-test.jl:162]: underscores, not minus

**Comment by wbhart** [comment]
Looks good overall. Just a few niggles.

**Comment by fredrik-johansson** [comment]
> In the docstring and in the docs, you are using `*a*` quite a alot. I think all of these should be `` `a` ``? I assume they should be formated as code and not italicized?

It's just more convenient to type. I will fix this for consistency.

File: src/calcium/ca.jl (hunk 3 above)
Comment (fredrik-johansson) [reply src/calcium/ca.jl:24]: Yeah, I copied this from another module and I was just not sure about the meaning.

File: src/calcium/CalciumTypes.jl (hunk 2 above)
Comment (fredrik-johansson) [reply src/calcium/CalciumTypes.jl:203]: Yeah, this is precisely the path that only gets called if Julia calls finalizers out of order, which only happens on some versions and on some machines. However, introducing a helper function should solve the coverage issue.

File: src/calcium/ca.jl (hunk 5 above)
Comment (fredrik-johansson) [reply src/calcium/ca.jl:44]: Yes, it's a nontrivial function to implement (well).

File: src/calcium/ca.jl (hunk 9 above)
Comment (fredrik-johansson) [reply src/calcium/ca.jl:1351]: Nonetheless, this check did help catch a bug. I'd rather keep it until the code is more battle-tested.

File: src/calcium/qqbar.jl (hunk 10 above)
Comment (fredrik-johansson) [reply src/calcium/qqbar.jl:785]: Yes, it's the usual way to define a named operator.

File: src/calcium/ca.jl (hunk 6 above)
Comment (fredrik-johansson) [reply src/calcium/ca.jl:184]: No idea what's happening behind the scenes, but the reverse operation works and is tested. I will add the reverse functions anyway to make sure it's fast.

File: src/calcium/ca.jl (hunk 7 above)
Comment (fredrik-johansson) [reply src/calcium/ca.jl:326]: Ok.

**Comment by fredrik-johansson** [comment]
@wbhart @thofma Niggles fixed, I think.

**Review by wbhart — APPROVED** [review]

**Pull request merged**